Event handler: introduce virtual desktop announcement switch handling

josephsl opened the pull request:
Note: this is take 3.1 of #5641 to see if a PR build can be built:

### Link to issue number:
Fixes #5641 

### Summary of the issue:
In Windows 10 and later, NVDA does not announce virtual desktop names wen they are opened, switched, and closed.

### Description of user facing changes
NVDA will announce names of virtual desktops when opening, switching, and closing them.

### Description of development approach
Event handler is extended to include a virtual desktop switch handler and an extension to execute event to catch a name change event coming from CSRSS.

### Testing strategy:
Manual testing (prerequisites: Windows 10 or later):

1. Press Control+Windows+D to open virtual desktops
2. Switch between virtual desktops.
3. Close a virtual desktop.

### Known issues with pull request:
In some cases, closing a virtual desktop wil announce the focused control first and then announce virtual desktop name, but this is due to focus event firing before name change event.

### Change log entries:
New features:

In Windows 10 May 2019 Update and later, NVDA can announe virtual desktop names when opening, changing, and closing them. (#5641)

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/devDocs/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

josephsl committed:
Event handler: update copyright header. Re #5641

josephsl committed:
Event handler/virtual desktops announcement: use a dedicated function to handle possible virtual desktop switches. Re #5641.

Original work by Jamie Teh (Mozilla): define a dedicated function to handle posible (and real) Windows 10/11 virtual desktop switche announcements. This function checks to make sure this is Widows 10 or later, and if yes, announces desktop name as defined in the event handler module, and clears this string once announcements are done. The actual event responsible for placing virtual desktop name in event handler is name change event from CSRSS (client/server runtime subsystem/Windows subsystem) process, specificlaly desktop object. Handling name change event will be done as part of handling gain focus event, specifically when doing pre-gain focus routine.

josephsl committed:
Event handler/virtual desktop switch announcements: detect name change event coming from CSRSS/desktop object and treat as a possible virtual desktop switch. Re #5641.

Original idea from Jamie Teh (Mozilla): in 'execute event' function, detect namechange coming from CSRSS/desktop object and treat it as a possible virtual desktop name change before handling focus events. If it turns out this is indeed a virtual desktop switch, set virtual desktop name, then queue handle possible virtual desktop switch routine (250 milliseconds) so desktop switches can be announced when opening/closing/switching virtual desktops while focused on the desktop.

josephsl committed:
Event handler/virtual desktop switch announcements: handle possible virtual desktop switches after handling new foreground event. Re #5641.

Credit: Jamie Teh (Mozilla): handle possible virtual desktop switch announcements after handling new foreground announcement in 'do pre gain focus' routine. This allows virtual desktop names to be announced between actual desktop switch and handling gain focus event, more noticeable when creating new virtual desktops.

josephsl committed:
Event handler/virtual desktop switch announcement: limit announcement to Windows 10 Version 1903 and later. Re #5641.

A note from Jamie Teh (Mozila): virtual desktop switch announcement works more effectively in Windows 10 Version 1903 (May 2019 Update) and later, therefore restrict virtual desktop announcement to that feature update and later.

josephsl committed:
Event handler/virtual desktop switch announcement and surrounding code: lint

josephsl committed:
Event handler/virtual desktop switch announcement: use formatted string literals when logging exceptions with execute event routine.

josephsl committed:
Event handler/virtual desktop switch announcement: add typing info for virtual desktop name field and handle possible desktop switch function. Re #5641

josephsl committed:
Event handler/virtual desktop switch announcement: use a dedicated variable to detect Windows 10 1903 and later. Re #5641.

In addition to placing winVersion module import at the top of the file, use a dedicated flag to indicate support for virtual desktop switch announcements in Windows 10 Version 1903 (May 2019 Update) and later.

josephsl committed:
Event handler: move virtual desktop constants to the top of the file.

Reported by several testers: in some cases, NVDA fails to start due to type hint scope issue if a constant is placed between functions. Therefore, move virtual desktops handling to top fo the file next to other constants and data structures.

josephsl committed:
Event handler: lint (Flake8 ET126)

josephsl commented:
CC @SeanBudd, @JCSTeh, @michaelDCurran

seanbudd commented on source/eventHandler.py:
```diff
@@ -37,6 +38,12 @@
 
+
+# Handle virtual desktop switch announcements in Windows 10 and later
+virtualDesktopName: Optional[str] = None
+canAnnounceVirtualDesktopNames: bool = winVersion.getWinVer() >= winVersion.WIN10_1903
```
are these expected to be consumed by add-ons? perhaps these should be private

seanbudd commented on source/eventHandler.py:
```diff
@@ -287,19 +294,45 @@ def executeEvent(
+		log.exception(f"error executing event: {eventName} on {obj} with extra args of {kwargs}")
+
+
+def handlePossibleDesktopNameChange() -> None:
```
same here - this should probably be private

seanbudd reviewed (commented):
Thanks @josephsl - the general approach here looks good

josephsl commented:
Hi, at least the function name (handle virtual desktop switch) is needed by Windows App Essentials add-on to determine NVDA Core patching determination. Others might become private, and I will take care of it in an upcoming commit. Thanks.

josephsl committed:
Event handler/virtual desktop announcements: make virtual desktop name and announcement check variables private. Re #5641.

Reviewed by Sean Budd (NV Access): make virtual desktop name and can nanouncement virtual desktop name variables private - at the moment no add-on appears to need this functionality but this can change if requested by add-on authors.

seanbudd commented:
Hi @josephsl - there are some strange artifacts in the diff here from other commits, including CLDR changes

josephsl commented:
Hi, when I did git pull origin master, it somehow did a rebase, and the CLDR commit artifacts came from git submodule update –init. Thanks.

CyrilleB79 commented:
> Hi, when I did git pull origin master, it somehow did a rebase, and the CLDR commit artifacts came from git submodule update –init. Thanks.

I do not know how to explain this. Maybe you have something strange in your git config when doing `git pull`.

Anyway, no only the CLDR submodule but also the symbol files appear in the modified files of this PR. This should be fixed please. Thanks.

josephsl commented:
Hi, I see. I’ll do something about this. Thanks

XLTechie commented:
I noticed the same thing happening on @josephsl commits a week or two back, but 
I think that PR was closed.

See the first answer here for what probably happened: 
https://stackoverflow.com/questions/9264314/git-commits-are-duplicated-in-the-same-branch-after-doing-a-rebase

TL;DR: you may have pulled because git told you to, but you should have force 
pushed.

josephsl committed:
Merge branch 'master' into i5641virtualDesktopAnnouncements

josephsl commented:
Should be resolved - did a force push as noted by a source from Luke's post above.

seanbudd reviewed (approved):
Thanks @josephsl

seanbudd committed:
Merge remote-tracking branch 'origin/master' into i5641virtualDesktopAnnouncements

seanbudd committed:
update changes
